### PR TITLE
Add metadata to keep track on who initiated a run

### PR DESCRIPTION
**Pull request by elfassy**
# Why

It is important for auditing purposes to know who initiated a maintenance task run. This PR adds an option to pass along run metadata that will be persisted in DB. 

part of https://hackdays.shopify.io/projects/16380

# What

- Adding a new optional `MaintenanceTasks.metadata` so users can set which method to call on their controller to get extra controller context (optional)
- Persist this data on the`maintenance_tasks_runs` table

```ruby
# config/initializers/maintenance_tasks.rb
MaintenanceTasks.parent_controller = "YourAdminController"

# NEW
MaintenanceTasks.metadata = -> do 
  { 
    user_email: current_user.email, 
    request_ip: request.ip,
  }
end 
```

```ruby
class YourAdminController < ActionController::Base
  def current_user
    ...
  end
end
```

### CODE REVIEW

File: db/migrate/20230622035228_add_user_email_to_runs.rb
Comment (elfassy) [reply db/migrate/20230622035228_add_user_email_to_runs.rb:7]: it's magic... in that they need to run `bin/rails maintenance_tasks:install:migrations` or manually as was done here https://github.com/Shopify/banking/pull/7880 😅 . This is a rough edge of engines gems

File: lib/maintenance_tasks.rb
Comment (etiennebarrie) [reply lib/maintenance_tasks.rb:92]: If we keep the option, I would definitely bike shed that. Rather than `user`, I'd rather something specific about the role they're playing, so maybe `initiator` or `creator`? And instead of `email`, we should be less specific because nothing forces it to be an email, since it's defined by the application. Implementors may use the field to store a `user id` instead.

File: test/application_system_test_case.rb
Comment (etiennebarrie) [inline test/application_system_test_case.rb:8]: We shouldn't need that either, we'll take care of the issue on main.

**Comment by nvasilevski** [comment]
>in a typical deployment, there is visibility/auditability on who writes a migration (and who reviews it) and who/when the migration was deployed to production. Here we're missing the information about who ran the task (apart from logs), so that might be a good argument in favor.

I think this is a good argument this is why I would be an favor of such feature but I find the `user_email` concept to be very limited or way too specific. I think we might be able to provide much more flexibility by not calling this concept as anything specific but instead build a way for an application to attach some "metadata" to the run

So the `user_email` column in the database could be a serializable field instead and `user_email_method` method could instead be a `proc` that knows how to populate the metadata field. On top of that we could add an empty partial that is being rendered on the UI so applications can override it to render certain parts of metadata on the UI

File: lib/maintenance_tasks.rb
Comment (nvasilevski) [reply lib/maintenance_tasks.rb:92]: I'm also leaning towards being less specific but was wondering if we should go even further and just allow some data to be stored in a serialized column. This way applications can decide how to call their actors and even decide to store more context about the actor in the field. A serialized column won't be searchable but I don't think this is something applications will need

**Comment by elfassy** [comment]
@nvasilevski I agree! I'll leave the UI part as a follow up PR, but happy to update this one with your recommendations

File: test/system/maintenance_tasks/tasks_test.rb
Comment (elfassy) [reply test/system/maintenance_tasks/tasks_test.rb:68]: will move this commit to a new PR, this test was invalid due to the `to_prepare` block being overwritten

File: test/application_system_test_case.rb
Comment (elfassy) [reply test/application_system_test_case.rb]: opened a PR here https://github.com/Shopify/maintenance_tasks/pull/860

File: test/test_helper.rb
Comment (elfassy) [reply test/test_helper.rb:12]: ex: https://github.com/Shopify/maintenance_tasks/actions/runs/5672921352/job/15373220595

**Review by robinbrandt — APPROVED** [review]

File: README.md
Comment (adrianna-chang-shopify) [inline README.md:839]:
```suggestion
or email of the user who performed the maintenance task.
```

File: lib/maintenance_tasks.rb
Comment (adrianna-chang-shopify) [inline lib/maintenance_tasks.rb:90]:
```suggestion
  #   @return [Proc] generates a hash containing the metadata to be stored on the Run
```

File: README.md
Comment (adrianna-chang-shopify) [inline README.md:838]:
```suggestion
Since this proc will be ran in the context of the `MaintenanceTasks.parent_controller`, it can be used to keep the id
```

**Review by adrianna-chang-shopify — APPROVED** [review]
This solution seems appropriately generic to me 👍 Let's wait on feedback from @nvasilevski and @etiennebarrie before merging though.

File: app/controllers/maintenance_tasks/runs_controller.rb
Comment (segiddins) [inline app/controllers/maintenance_tasks/runs_controller.rb:17]: The README states that the `metadata` proc will be exec'd in the scope of the controller, but this isn't doing that -- did you intend to `instance_exec(MaintenanceTasks.metadata) if MaintenanceTasks.metadata` ?

**Pull request merged**
